src/api/tx_submit.rs

```diff
@@ -29,3 +32,37 @@ pub async fn route(
 
     Ok(Json(response))
 }
+
+/// This function allows us to take both hex-encoded and raw bytes. It has
+/// to be a heuristic: if there are input bytes that are not `[0-9a-f]`,
+/// then it must be a binary string. Otherwise, we assume it’s hex encoded.
+///
+/// **Note**: there is a small probability that the user gave us a binary
+/// string that only _looked_ like a hex-encoded one, but it’s rare enough
+/// to ignore it.
+pub fn binary_or_hex_heuristic(xs: &[u8]) -> Vec<u8> {
+    let even_length = xs.len() % 2 == 0;
+    let contains_non_hex = xs.iter().any(|&x| !x.is_ascii_hexdigit());
+    if !even_length || contains_non_hex {
+        xs.to_vec()
+    } else {
+        hex::decode(xs).unwrap_or_else(|_| unreachable!())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::prelude::*;
+
+    proptest! {
+        #[test]
+        fn proptest_binary_or_hex_heuristic(binary in prop::collection::vec(any::<u8>(), 0..=128)) {
```

Comment:
And it's a good and simple opportunity to introduce `proptest` (like QuickCheck in Haskell). We already use property testing of `ApplyTxErr`, but we set it up ourselves. This is a popular library for Rust-only tests.